internal/daemon/api_files.go

```diff
@@ -0,0 +1,594 @@
+		info, ok := infos[p]
+		if !ok {
+			return statusBadRequest("no metadata for path %q", p)
+		}
+		errors[p] = writeFile(info, part)
+		part.Close()
+	}
+
+	// Build list of results with any errors.
+	result := make([]fileResult, len(payload.Files))
+	var firstErr error
+	for i, file := range payload.Files {
+		err, ok := errors[file.Path]
+		if !ok {
+			// Ensure we wrote all the files in the metadata.
+			err = fmt.Errorf("no file content for path %q", file.Path)
+		}
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		result[i] = fileResult{
+			Path:  file.Path,
+			Error: fileErrorToResult(err),
+		}
+	}
+	return syncResponseWithError(result, firstErr)
+}
+
+func writeFile(item writeFilesItem, source io.Reader) error {
+	if !path.IsAbs(item.Path) {
+		return fmt.Errorf("paths must be absolute (%q is not)", item.Path)
+	}
+
+	// Create parent directory if needed
+	if item.MakeDirs {
+		err := os.MkdirAll(path.Dir(item.Path), 0o755)
+		if err != nil {
+			return fmt.Errorf("error creating directory: %w", err)
+		}
+	}
+
+	// Create file and write contents.
+	perm, err := parsePermissions(item.Permissions, 0o644)
+	if err != nil {
+		return err
+	}
+	f, err := os.OpenFile(item.Path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return fmt.Errorf("error creating file: %w", err)
+	}
+	_, err = io.Copy(f, source)
+	if err != nil {
+		_ = f.Close()
+		return fmt.Errorf("error writing file: %w", err)
+	}
+	err = f.Close()
+	if err != nil {
+		return fmt.Errorf("error closing file: %w", err)
+	}
+
+	// Update user and group if necessary.
+	err = updateUserGroup(item.Path, item.UserID, item.User, item.GroupID, item.Group)
+	if err != nil {
+		return fmt.Errorf("error setting user and group: %w", err)
+	}
+	return nil
+}
+
+func parsePermissions(permissions string, defaultMode os.FileMode) (os.FileMode, error) {
+	if permissions == "" {
+		return defaultMode, nil
+	}
+	perm, err := strconv.ParseUint(permissions, 8, 32)
+	if err != nil || len(permissions) != 3 {
+		return 0, fmt.Errorf("permissions must be a 3-digit octal string, not %q", permissions)
+	}
+	return os.FileMode(perm), nil
+}
+
+type makeDirsItem struct {
+	Path        string `json:"path"`
+	MakeParents bool   `json:"make-parents"`
+	Permissions string `json:"permissions"`
+	UserID      int    `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     int    `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+func makeDirs(dirs []makeDirsItem) Response {
+	result := make([]fileResult, len(dirs))
+	var firstErr error
+	for i, dir := range dirs {
+		err := makeDir(dir)
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		result[i] = fileResult{
+			Path:  dir.Path,
+			Error: fileErrorToResult(err),
+		}
+	}
+	return syncResponseWithError(result, firstErr)
+}
+
+func makeDir(dir makeDirsItem) error {
+	if !path.IsAbs(dir.Path) {
+		return fmt.Errorf("paths must be absolute (%q is not)", dir.Path)
```

Comment:
not sure if we want a msg template const or helper func for this common error

Reply:
Good call, done.